sidecar/src/log.rs

```diff
@@ -0,0 +1,324 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present Datadog, Inc.
+
+use crate::config;
+use chrono::{DateTime, Utc};
+use lazy_static::lazy_static;
+use priority_queue::PriorityQueue;
+use std::cmp::max;
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::ops::Sub;
+use std::path::PathBuf;
+use std::sync::{Mutex, RwLock};
+use std::time::{Duration, Instant, SystemTime};
+use std::{env, io};
+use tracing::level_filters::LevelFilter;
+use tracing::span::{Attributes, Record};
+use tracing::subscriber::Interest;
+use tracing::{Event, Id, Metadata, Subscriber};
+use tracing_log::LogTracer;
+use tracing_subscriber::fmt::format::Writer;
+use tracing_subscriber::fmt::{FmtContext, FormatEvent, FormatFields, MakeWriter};
+use tracing_subscriber::layer::{Context, Filter, SubscriberExt};
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::util::SubscriberInitExt;
+use tracing_subscriber::{EnvFilter, Layer};
+
+fn create_logfile(path: &PathBuf) -> anyhow::Result<std::fs::File> {
+    let log_file = std::fs::File::options()
+        .create(true)
+        .truncate(false)
+        .write(true)
+        .append(true)
+        .open(path)?;
+    Ok(log_file)
+}
+
+pub struct TemporarilyRetainedMap<K, V>
```

Comment:
Maybe the code and the logic in this file is self explanatory to everyone else, but I would welcome some comments outlining how things fit together. Especially the interaction between this code and the code in `tracing` that calls back into this code to compute the log levels.

Also, are there any tests for this functionality?

Reply:
I agree, let me add a couple comments.

The sidecar code is mostly tested through tests from the dd-trace-php repository. I haven't written self-contained tests for the sidecar yet.

Reply:
I would really like something as complex as a reference counting data structure (with the potential to leak memory) to have some unit tests.